{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4e9e2529-c2f1-46a6-8399-a89c0ca64dfb",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " # Revision de Python\n",
    " En esta notebook vamos a repasar algunos conceptos basicos de Python\n",
    " con algunos ejercicios.\n",
    "\n",
    " Más adelante trabajaremos con algunas librerias de comunes de Python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c99b67bd-0f88-4b41-823d-3b0254babd76",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ## Ejercicios de Python\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contar nucleótidos\n",
    "\n",
    "Dada una cadena de ADN, cuenta cuántas veces aparece cada nucleótido (A, T, C, G).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5 A, 4 T, 3 C, 2 G\n",
    "seq = \"AAAAATTTTCCCGG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_nucleotides_1(seq):\n",
    "    \"\"\"Count the number of each nucleotide in a sequence.\"\"\"\n",
    "    n_a = 0\n",
    "    n_c = 0\n",
    "    n_g = 0\n",
    "    n_t = 0\n",
    "    for nucleotide in seq:\n",
    "        if nucleotide == \"A\":\n",
    "            n_a += 1\n",
    "        elif nucleotide == \"C\":\n",
    "            n_c += 1\n",
    "        elif nucleotide == \"G\":\n",
    "            n_g += 1\n",
    "        elif nucleotide == \"T\":\n",
    "            n_t += 1\n",
    "    return n_a, n_c, n_g, n_t\n",
    "\n",
    "assert count_nucleotides_1(seq) == (5, 3, 2, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_nucleotides_2(seq):\n",
    "    \"\"\"Count the number of each nucleotide in a sequence.\"\"\"\n",
    "    counts = {}\n",
    "    for nucleotide in seq:\n",
    "        if nucleotide in counts:\n",
    "            counts[nucleotide] += 1\n",
    "        else:\n",
    "            counts[nucleotide] = 1\n",
    "    return counts.get(\"A\", 0), counts.get(\"C\", 0), counts.get(\"G\", 0), counts.get(\"T\", 0)\n",
    "\n",
    "assert count_nucleotides_2(seq) == (5, 3, 2, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "def count_nucleotides_3(seq):\n",
    "    \"\"\"Count the number of each nucleotide in a sequence.\"\"\"\n",
    "    counts = defaultdict(int)\n",
    "    for nucleotide in seq:\n",
    "        counts[nucleotide] += 1\n",
    "    return counts.get(\"A\", 0), counts.get(\"C\", 0), counts.get(\"G\", 0), counts.get(\"T\", 0)\n",
    "\n",
    "assert count_nucleotides_3(seq) == (5, 3, 2, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "def count_nucleotides_4(seq):\n",
    "    \"\"\"Count the number of each nucleotide in a sequence.\"\"\"\n",
    "    counts = Counter(seq)\n",
    "    return counts.get(\"A\", 0), counts.get(\"C\", 0), counts.get(\"G\", 0), counts.get(\"T\", 0)\n",
    "\n",
    "assert count_nucleotides_4(seq) == (5, 3, 2, 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Transcripción de ADN a ARN\n",
    "\n",
    "Convierte una cadena de ADN en ARN reemplazando todas las T por U.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def trancript_1(adn: str) -> str:\n",
    "    \"\"\"Convert a DNA sequence to an RNA sequence.\"\"\"\n",
    "    new_seq = []\n",
    "    for nucleotide in adn:\n",
    "        if nucleotide == \"T\":\n",
    "            new_seq.append(\"U\")\n",
    "        else:\n",
    "            new_seq.append(nucleotide)\n",
    "    return \"\".join(new_seq)\n",
    "\n",
    "assert trancript_1(\"ATCGTACG\") == \"AUCGUACG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trancript_2(adn: str) -> str:\n",
    "    \"\"\"Convert a DNA sequence to an RNA sequence.\"\"\"\n",
    "    return \"\".join([\"U\" if nucleotide == \"T\" else nucleotide for nucleotide in adn])\n",
    "assert trancript_2(\"ATCGTACG\") == \"AUCGUACG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trancript_3(adn: str) -> str:\n",
    "    \"\"\"Convert a DNA sequence to an RNA sequence.\"\"\"\n",
    "    return adn.replace(\"T\", \"U\")\n",
    "assert trancript_3(\"ATCGTACG\") == \"AUCGUACG\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Secuencia reversa complementaria\n",
    "\n",
    "Devuelve la cadena reversa complementaria de una secuencia de ADN.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def reverse_complement_1(seq: str) -> str:\n",
    "    \"\"\"Return the reverse complement of a DNA sequence.\"\"\"\n",
    "    new_seq = []\n",
    "    for nucleotide in seq:\n",
    "        if nucleotide == \"A\":\n",
    "            new_seq.append(\"T\")\n",
    "        elif nucleotide == \"T\":\n",
    "            new_seq.append(\"A\")\n",
    "        elif nucleotide == \"C\":\n",
    "            new_seq.append(\"G\")\n",
    "        elif nucleotide == \"G\":\n",
    "            new_seq.append(\"C\")\n",
    "    return \"\".join(reversed(new_seq))\n",
    "\n",
    "assert reverse_complement_1(\"AAATTTCCCGGG\") == \"CCCGGGAAATTT\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def reverse_complement_2(seq: str) -> str:\n",
    "    \"\"\"Return the reverse complement of a DNA sequence.\"\"\"\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    return \"\".join(complement[nucleotide] for nucleotide in reversed(seq))\n",
    "\n",
    "assert reverse_complement_2(\"AAATTTCCCGGG\") == \"CCCGGGAAATTT\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### GC content\n",
    "\n",
    "Calcula el porcentaje de contenido GC de una secuencia de ADN.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gc_content(seq: str) -> float:\n",
    "    \"\"\"Return the GC content of a DNA sequence.\"\"\"\n",
    "    n_gc = 0\n",
    "    for nucleotide in seq:\n",
    "        if nucleotide == \"G\" or nucleotide == \"C\":\n",
    "            n_gc += 1\n",
    "    return n_gc / len(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Palíndromos en ADN\n",
    "\n",
    "Verifica si una cadena de ADN es palindrómica (igual a su reversa complementaria).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def is_palindrome(seq1:str) -> bool:\n",
    "    \"\"\"Check if two sequences are palindromes of each other.\"\"\"\n",
    "    return seq1 == reverse_complement_2(seq1)\n",
    "\n",
    "assert is_palindrome(\"AAACCCGGGTTT\") == True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Traducción de ADN a proteína\n",
    "\n",
    "Dada una cadena de ADN y la tabla de codones, tradúcela en su secuencia de aminoácidos.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_table = {\n",
    "    # Alanine\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    # Isoleucine\n",
    "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\",\n",
    "    # Arginine\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    # Leucine\n",
    "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\", \"TTA\": \"L\", \"TTG\": \"L\",\n",
    "    # Asparagine\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
    "    # Lysine\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    # Aspartic Acid\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
    "    # Methionine\n",
    "    \"ATG\": \"M\",\n",
    "    # Phenylalanine\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
    "    # Cysteine\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
    "    # Proline\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    # Glutamine\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    # Serine\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
    "    # Glutamic Acid\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    # Threonine\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    # Tryptophan\n",
    "    \"TGG\": \"W\",\n",
    "    # Glycine\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "    # Tyrosine\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
    "    # Histidine\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
    "    # Valine\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    # STOP codons\n",
    "    \"TAA\": \"STOP\", \"TGA\": \"STOP\", \"TAG\": \"STOP\"\n",
    "}\n",
    "\n",
    "assert len(codon_table) == 64\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def translate(seq: str, codon_table: dict[str, str]) -> str:\n",
    "    \"\"\"Translate a DNA sequence into a protein sequence.\"\"\"\n",
    "    protein = []\n",
    "    for i in range(0, len(seq), 3):\n",
    "        codon = seq[i:i + 3]\n",
    "        if len(codon) < 3:\n",
    "            break\n",
    "        amino_acid = codon_table.get(codon, \"X\")\n",
    "        if amino_acid == \"STOP\":\n",
    "            break\n",
    "        protein.append(amino_acid)\n",
    "    return \"\".join(protein)\n",
    "\n",
    "assert translate(\"ATGTTTAAAGGG\", codon_table) == \"MFKG\"\n",
    "assert translate(\"ATGTTTAAAGGGA\", codon_table) == \"MFKG\"\n",
    "assert translate(\"ATGTTTAAAGGGAA\", codon_table) == \"MFKG\"\n",
    "assert translate(\"ATGTTTAAAGGGTAA\", codon_table) == \"MFKG\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Buscar codones de inicio y stop\n",
    "\n",
    "Encuentra todas las posiciones donde aparece un codón de inicio (ATG) y su correspondiente codón stop (TAA, TAG, TGA).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scan_atg(seq: str) -> list[int]:\n",
    "    \"\"\"Scan a DNA sequence for the start codon ATG.\"\"\"\n",
    "    start_codons = []\n",
    "    for i in range(len(seq) - 2):\n",
    "        if seq[i:i + 3] == \"ATG\":\n",
    "            start_codons.append(i)\n",
    "    return start_codons\n",
    "\n",
    "def scan_stop(seq: str) -> list[int]:\n",
    "    \"\"\"Scan a DNA sequence for the stop codon.\"\"\"\n",
    "    start_codons = []\n",
    "    for i in range(len(seq) - 2):\n",
    "        codon = seq[i:i + 3]\n",
    "        if codon == \"TAA\" or codon == \"TAG\" or codon == \"TGA\":\n",
    "            start_codons.append(i)\n",
    "    return start_codons\n",
    "\n",
    "def find_atg_stop(seq: str) -> list[tuple[int, int]]:\n",
    "    \"\"\"Find all start and stop codons in a DNA sequence.\"\"\"\n",
    "    start_codons = scan_atg(seq)\n",
    "    stop_codons = scan_stop(seq)\n",
    "    atg_stop = []\n",
    "    for start in start_codons:\n",
    "        for stop in stop_codons:\n",
    "            if stop > start and (stop - start) % 3 == 0:\n",
    "                atg_stop.append((start, stop))\n",
    "                break\n",
    "    return atg_stop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Conteo de motivos\n",
    "\n",
    "Dada una secuencia y un motivo (subsecuencia), cuenta cuántas veces aparece (posiciones solapadas incluidas).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Generador de FASTA\n",
    "\n",
    "Crea una función que reciba un identificador y una secuencia, y genere una entrada en formato FASTA.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Parseador de FASTA\n",
    "\n",
    "Lee un archivo FASTA y devuelve un diccionario con identificadores como claves y secuencias como valores.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Distancia de Hamming\n",
    "\n",
    "Calcula la distancia de Hamming entre dos secuencias de igual longitud.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Subregiones palindrómicas\n",
    "\n",
    "Dada una secuencia encuentra las subregiones palindrómicas de al menos 5 bases.\n",
    "Las regiones pueden solaparse, pero no estar una contenida dentro de otra. En\n",
    "estos casos traer la de mayor tamaño.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    " ### Datos de Disprot\n",
    " Leer los datos del archivo classes/data/DisProt.tsv\n",
    " 1. Contar la cantidad de regiones para cada accession.\n",
    " 2. En aquellos casos donde hay más de una región, contar las que tienen\n",
    "    posiciones de principio y fin diferentes.\n",
    " 3  Estimar la longitud total de las proteínas en el archivo.\n",
    " 4. Contar los regiones que tienen una longitud mayor a 10 aminoácidos.\n",
    " 5. Generar una nueva tabla Las filas correposden\n",
    "    a las proteinas y las columnas son: accession, y luego tiene una columna\n",
    "    por cada valor de term_namespace. En cada celda se debe contar la cantidad\n",
    "    de veces que aparece el valor de term_namespace en la proteina."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f97607a-a88e-49d5-a682-78d563fc7be7",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Datos de taxonomy\n",
    " Leer los datos del archivo classes/data/nodes.csv.\n",
    " El archivo tiene las siguientes columnas: tax_id, parent_tax_id, rank.\n",
    " 1. Verificar que los datos forma un grafo aciclico con una unica\n",
    "    componente.\n",
    " 2. Contar la cantidad de nodos hoja.\n",
    " 3. Contar la cantidad de nodos internos.\n",
    " 4. Contar la profundidad de cada hoja.\n",
    " 5. Calcular el ancestro comun mas cercano entre dos nodos.\n",
    " 6. Calcular la distancia entre dos nodos.\n",
    " 7. Contar la cantidad de nodos que tienen un ranking inferior a especie."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "625015dc-6d72-4851-abff-c1a36925b528",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Simulación de secuencias de aminoácidos\n",
    "1. A partir de una secuencia de aminoácidos, simular un proceso de evolución\n",
    "   de la secuencia. Para esto se debe definir una matriz de probabilidades de\n",
    "   sustitución de aminoácidos.\n",
    "   - La evolución de la secuencia se puede simular de a pasos de tiempo hasta\n",
    "     un tiempo total de evolución (1000 años).\n",
    "   - Cada paso de tiempo es elegido al azar entre 25 y 200 años.\n",
    "   - En cada paso de tiempo, se generan dos secuencias hijas a partir de cada\n",
    "     secuencia padre.\n",
    "   - Para cada secuencia hija, todos los aminoácidos pueden ser sustituidos\n",
    "     por otro aminoácido con una tasa dada por la matriz de\n",
    "     sustitución y el tiempo de evolución (substitution_rates.csv).\n",
    "   - La composicion del aminoácidos de la secuencia inicial debe corresponderse\n",
    "     con la composición de aminoácidos general observada en la en naturaleza\n",
    "     (ver tabla de amino_frq.csv). Asuma que las probabilidades de sustitución\n",
    "     son independientes de la posición en la secuencia y que la tasa\n",
    "     de sustitución esta dada en sustituciones cada mil años.\n",
    " 2. Calcular el porcentaje de identidad entre las secuencias hijas y la\n",
    "    secuencia padre.\n",
    " 3. Hacer lo mismo asumiendo que la puede haber inserciones y deleciones en la\n",
    "    secuencia. En este caso, la probabilidad de inserción y deleción es\n",
    "    independiente de la posición en la secuencia y de la longitud de la\n",
    "    secuencia. La probabilidad de inserción es 0.0001 cada 100 años y la\n",
    "    probabilidad de deleción es 0.0001 cada 100 años. La cantidad de aminoácidos\n",
    "    insertados o eliminados es un número al azar entre 1 y 5."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "574af66a-7c72-4749-8cd3-557bd8cb42fb",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ## Pandas\n",
    " - Pandas es una libreria que permite el trabajar con datos tabulares.\n",
    " - Permite leer y escribir datos en diferentes formatos.\n",
    " - Permite hacer operaciones de filtrado, agrupamiento, y transformación de\n",
    "   datos."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63abbee1-0328-4096-ae13-7f639ae01641",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " Como ejemplo, vamos a leer los datos de DisProt."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b5326410-5717-4ff2-baf7-b6d5e40822a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "disprot = pd.read_csv('data/DisProt.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dc5fa69-8cba-4eb7-947d-d128b6db05f3",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Información de general de los datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3d3f3a0b-eae6-4f36-86b4-d12723420526",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>name</th>\n",
       "      <th>disorder_content</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncbi_taxon_id</th>\n",
       "      <th>disprot_id</th>\n",
       "      <th>region_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>term_namespace</th>\n",
       "      <th>term</th>\n",
       "      <th>term_name</th>\n",
       "      <th>ec</th>\n",
       "      <th>ec_name</th>\n",
       "      <th>reference</th>\n",
       "      <th>region_sequence</th>\n",
       "      <th>confidence</th>\n",
       "      <th>obsolete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>POU domain class 2-associating factor 1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00008</td>\n",
       "      <td>DP00008r013</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0006210</td>\n",
       "      <td>small-angle X-ray scattering evidence used in ...</td>\n",
       "      <td>pmid:11380252</td>\n",
       "      <td>MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>POU domain class 2-associating factor 1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00008</td>\n",
       "      <td>DP00008r015</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0006232</td>\n",
       "      <td>differential scanning calorimetry evidence use...</td>\n",
       "      <td>pmid:10329190</td>\n",
       "      <td>MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>POU domain class 2-associating factor 1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00008</td>\n",
       "      <td>DP00008r020</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0006204</td>\n",
       "      <td>far-UV circular dichroism evidence used in man...</td>\n",
       "      <td>pmid:10329190</td>\n",
       "      <td>MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>POU domain class 2-associating factor 1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00008</td>\n",
       "      <td>DP00008r023</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>Structural transition</td>\n",
       "      <td>IDPO:00050</td>\n",
       "      <td>disorder to order</td>\n",
       "      <td>ECO:0006204</td>\n",
       "      <td>far-UV circular dichroism evidence used in man...</td>\n",
       "      <td>pmid:10329190</td>\n",
       "      <td>MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>POU domain class 2-associating factor 1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00008</td>\n",
       "      <td>DP00008r024</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0005647</td>\n",
       "      <td>isothermal titration calorimetry evidence used...</td>\n",
       "      <td>pmid:10329190</td>\n",
       "      <td>MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      acc                                     name  disorder_content  \\\n",
       "0  Q64693  POU domain class 2-associating factor 1             100.0   \n",
       "1  Q64693  POU domain class 2-associating factor 1             100.0   \n",
       "2  Q64693  POU domain class 2-associating factor 1             100.0   \n",
       "3  Q64693  POU domain class 2-associating factor 1             100.0   \n",
       "4  Q64693  POU domain class 2-associating factor 1             100.0   \n",
       "\n",
       "       organism  ncbi_taxon_id disprot_id    region_id  start  end  \\\n",
       "0  Mus musculus          10090    DP00008  DP00008r013      1  256   \n",
       "1  Mus musculus          10090    DP00008  DP00008r015      1  256   \n",
       "2  Mus musculus          10090    DP00008  DP00008r020      1  256   \n",
       "3  Mus musculus          10090    DP00008  DP00008r023      1   65   \n",
       "4  Mus musculus          10090    DP00008  DP00008r024      1  256   \n",
       "\n",
       "          term_namespace        term          term_name           ec  \\\n",
       "0       Structural state  IDPO:00076           disorder  ECO:0006210   \n",
       "1       Structural state  IDPO:00076           disorder  ECO:0006232   \n",
       "2       Structural state  IDPO:00076           disorder  ECO:0006204   \n",
       "3  Structural transition  IDPO:00050  disorder to order  ECO:0006204   \n",
       "4     Molecular function  GO:0005515    protein binding  ECO:0005647   \n",
       "\n",
       "                                             ec_name      reference  \\\n",
       "0  small-angle X-ray scattering evidence used in ...  pmid:11380252   \n",
       "1  differential scanning calorimetry evidence use...  pmid:10329190   \n",
       "2  far-UV circular dichroism evidence used in man...  pmid:10329190   \n",
       "3  far-UV circular dichroism evidence used in man...  pmid:10329190   \n",
       "4  isothermal titration calorimetry evidence used...  pmid:10329190   \n",
       "\n",
       "                                     region_sequence  confidence  obsolete  \n",
       "0  MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...         NaN       NaN  \n",
       "1  MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...         NaN       NaN  \n",
       "2  MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...         NaN       NaN  \n",
       "3  MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...         NaN       NaN  \n",
       "4  MLWQKSTAPEQAPAPPRPYQGVRVKEPVKELLRRKRGHTSVGAAGP...         NaN       NaN  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disprot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52fa9817-3b6f-49af-8c87-58f410a7ea35",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Mostrar las columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a0a6c4b7-f67e-4eb5-b88a-ff9deec28188",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['acc', 'name', 'disorder_content', 'organism', 'ncbi_taxon_id',\n",
       "       'disprot_id', 'region_id', 'start', 'end', 'term_namespace', 'term',\n",
       "       'term_name', 'ec', 'ec_name', 'reference', 'region_sequence',\n",
       "       'confidence', 'obsolete'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disprot.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce1a9a21-9a2c-46a6-b47b-d0acdf7f5a2e",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Mostrar los primeros registros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e45dc617-8321-466d-8b19-f23794d6212f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "disprot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81cb03cb-17d7-4ca0-a95d-7e4aef2dee12",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Subsettear los datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d7cb126-e3d7-432b-8b3c-46223d033641",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>disorder_content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q64693</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      acc  disorder_content\n",
       "0  Q64693             100.0\n",
       "1  Q64693             100.0\n",
       "2  Q64693             100.0\n",
       "3  Q64693             100.0\n",
       "4  Q64693             100.0"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# disprot['acc']\n",
    "# Devuelve una pd.Serie\n",
    "\n",
    "# type(disprot.acc)\n",
    "# Devuelve una pd.Serie\n",
    "\n",
    "# disprot.filter(['acc'])\n",
    "# Devuelve un pd.DataFrame\n",
    "\n",
    "# disprot[['acc']]\n",
    "# Devuelve un pd.DataFrame\n",
    "\n",
    "# disprot[['acc', 'term_namespace']]\n",
    "\n",
    "# disprot.loc[0:5, ['acc', 'term_namespace']]\n",
    "# Aqui 0:5 hace referencia a los indices de las filas, no a la posicion de las\n",
    "# filas, aunque en este caso coinciden.\n",
    "\n",
    "# disprot.iloc[0:5, [0, 2]]\n",
    "# Aqui 0:5 hace referencia a los posiciones de las filas, no al indice de las\n",
    "# filas, aunque en este caso coinciden."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ef9be83-4f0d-4c2a-ab91-82368467880b",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Filtrar los datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cb623418-3b2c-42b7-8bfe-f12baf420812",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>name</th>\n",
       "      <th>disorder_content</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncbi_taxon_id</th>\n",
       "      <th>disprot_id</th>\n",
       "      <th>region_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>term_namespace</th>\n",
       "      <th>term</th>\n",
       "      <th>term_name</th>\n",
       "      <th>ec</th>\n",
       "      <th>ec_name</th>\n",
       "      <th>reference</th>\n",
       "      <th>region_sequence</th>\n",
       "      <th>confidence</th>\n",
       "      <th>obsolete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r001</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0006165</td>\n",
       "      <td>nuclear magnetic resonance spectroscopy eviden...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r002</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0060090</td>\n",
       "      <td>molecular adaptor activity</td>\n",
       "      <td>ECO:0006165</td>\n",
       "      <td>nuclear magnetic resonance spectroscopy eviden...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r003</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0006165</td>\n",
       "      <td>nuclear magnetic resonance spectroscopy eviden...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r004</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0007680</td>\n",
       "      <td>chromatography evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r005</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0060090</td>\n",
       "      <td>molecular adaptor activity</td>\n",
       "      <td>ECO:0007680</td>\n",
       "      <td>chromatography evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r006</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0007680</td>\n",
       "      <td>chromatography evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r007</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0007691</td>\n",
       "      <td>cleavage assay evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r008</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0060090</td>\n",
       "      <td>molecular adaptor activity</td>\n",
       "      <td>ECO:0007691</td>\n",
       "      <td>cleavage assay evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r009</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0007691</td>\n",
       "      <td>cleavage assay evidence used in manual assertion</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r010</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0006204</td>\n",
       "      <td>far-UV circular dichroism evidence used in man...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r011</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0060090</td>\n",
       "      <td>molecular adaptor activity</td>\n",
       "      <td>ECO:0006204</td>\n",
       "      <td>far-UV circular dichroism evidence used in man...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r012</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0006204</td>\n",
       "      <td>far-UV circular dichroism evidence used in man...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r013</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Structural state</td>\n",
       "      <td>IDPO:00076</td>\n",
       "      <td>disorder</td>\n",
       "      <td>ECO:0007064</td>\n",
       "      <td>dynamic light scattering assay evidence used i...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r014</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0060090</td>\n",
       "      <td>molecular adaptor activity</td>\n",
       "      <td>ECO:0007064</td>\n",
       "      <td>dynamic light scattering assay evidence used i...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>Q91V27</td>\n",
       "      <td>Melanophilin</td>\n",
       "      <td>43.56</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>DP00541</td>\n",
       "      <td>DP00541r015</td>\n",
       "      <td>147</td>\n",
       "      <td>403</td>\n",
       "      <td>Molecular function</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>ECO:0007064</td>\n",
       "      <td>dynamic light scattering assay evidence used i...</td>\n",
       "      <td>pmid:17513864</td>\n",
       "      <td>GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        acc          name  disorder_content      organism  ncbi_taxon_id  \\\n",
       "165  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "166  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "167  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "168  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "169  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "170  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "171  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "172  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "173  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "174  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "175  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "176  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "177  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "178  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "179  Q91V27  Melanophilin             43.56  Mus musculus          10090   \n",
       "\n",
       "    disprot_id    region_id  start  end      term_namespace        term  \\\n",
       "165    DP00541  DP00541r001    147  403    Structural state  IDPO:00076   \n",
       "166    DP00541  DP00541r002    147  403  Molecular function  GO:0060090   \n",
       "167    DP00541  DP00541r003    147  403  Molecular function  GO:0005515   \n",
       "168    DP00541  DP00541r004    147  403    Structural state  IDPO:00076   \n",
       "169    DP00541  DP00541r005    147  403  Molecular function  GO:0060090   \n",
       "170    DP00541  DP00541r006    147  403  Molecular function  GO:0005515   \n",
       "171    DP00541  DP00541r007    147  403    Structural state  IDPO:00076   \n",
       "172    DP00541  DP00541r008    147  403  Molecular function  GO:0060090   \n",
       "173    DP00541  DP00541r009    147  403  Molecular function  GO:0005515   \n",
       "174    DP00541  DP00541r010    147  403    Structural state  IDPO:00076   \n",
       "175    DP00541  DP00541r011    147  403  Molecular function  GO:0060090   \n",
       "176    DP00541  DP00541r012    147  403  Molecular function  GO:0005515   \n",
       "177    DP00541  DP00541r013    147  403    Structural state  IDPO:00076   \n",
       "178    DP00541  DP00541r014    147  403  Molecular function  GO:0060090   \n",
       "179    DP00541  DP00541r015    147  403  Molecular function  GO:0005515   \n",
       "\n",
       "                      term_name           ec  \\\n",
       "165                    disorder  ECO:0006165   \n",
       "166  molecular adaptor activity  ECO:0006165   \n",
       "167             protein binding  ECO:0006165   \n",
       "168                    disorder  ECO:0007680   \n",
       "169  molecular adaptor activity  ECO:0007680   \n",
       "170             protein binding  ECO:0007680   \n",
       "171                    disorder  ECO:0007691   \n",
       "172  molecular adaptor activity  ECO:0007691   \n",
       "173             protein binding  ECO:0007691   \n",
       "174                    disorder  ECO:0006204   \n",
       "175  molecular adaptor activity  ECO:0006204   \n",
       "176             protein binding  ECO:0006204   \n",
       "177                    disorder  ECO:0007064   \n",
       "178  molecular adaptor activity  ECO:0007064   \n",
       "179             protein binding  ECO:0007064   \n",
       "\n",
       "                                               ec_name      reference  \\\n",
       "165  nuclear magnetic resonance spectroscopy eviden...  pmid:17513864   \n",
       "166  nuclear magnetic resonance spectroscopy eviden...  pmid:17513864   \n",
       "167  nuclear magnetic resonance spectroscopy eviden...  pmid:17513864   \n",
       "168   chromatography evidence used in manual assertion  pmid:17513864   \n",
       "169   chromatography evidence used in manual assertion  pmid:17513864   \n",
       "170   chromatography evidence used in manual assertion  pmid:17513864   \n",
       "171   cleavage assay evidence used in manual assertion  pmid:17513864   \n",
       "172   cleavage assay evidence used in manual assertion  pmid:17513864   \n",
       "173   cleavage assay evidence used in manual assertion  pmid:17513864   \n",
       "174  far-UV circular dichroism evidence used in man...  pmid:17513864   \n",
       "175  far-UV circular dichroism evidence used in man...  pmid:17513864   \n",
       "176  far-UV circular dichroism evidence used in man...  pmid:17513864   \n",
       "177  dynamic light scattering assay evidence used i...  pmid:17513864   \n",
       "178  dynamic light scattering assay evidence used i...  pmid:17513864   \n",
       "179  dynamic light scattering assay evidence used i...  pmid:17513864   \n",
       "\n",
       "                                       region_sequence  confidence  obsolete  \n",
       "165  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "166  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "167  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "168  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "169  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "170  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "171  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "172  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "173  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "174  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "175  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "176  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "177  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "178  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  \n",
       "179  GGGGSEPSLEEGNGDSEQTDEDGDLDTEARDQPLNSKKKKRLLSFR...         NaN       NaN  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# disprot.query('disorder_content > 50')\n",
    "\n",
    "# disprot.loc[disprot.disorder_content>50, :]\n",
    "\n",
    "# disprot.iloc[disprot.disorder_content>50, :] # Error\n",
    "\n",
    "upper = 60\n",
    "lower = 40\n",
    "# disprot.query('disorder_content > @lower and disorder_content < @upper')\n",
    "disprot.query('acc == \"Q91V27\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca2a3c5-b06e-4c3b-a597-13b8304f4b12",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Transformar los datos / Agregar columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "21420d03-3728-4227-9db3-d6b451425d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# disprot.assign(disorder_content_norm = disprot.disorder_content / 100)\n",
    "# No es in-place\n",
    "\n",
    "disprot['disorder_content_norm'] = disprot.disorder_content / 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53c6077-89bc-4aa8-83a5-bed4bcbd1b68",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Transformar los datos / Apply"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "9a78fbbb-c3b5-4de1-bf8c-fa152edbd316",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      256.0\n",
       "1      256.0\n",
       "2      256.0\n",
       "3      100.0\n",
       "4      256.0\n",
       "       ...  \n",
       "742     10.0\n",
       "743     60.0\n",
       "744    142.0\n",
       "745    220.0\n",
       "746     73.0\n",
       "Length: 747, dtype: float64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Applicar la misma función a todas las columnas o filas\n",
    "\n",
    "(\n",
    "  disprot\n",
    "    .filter(['disorder_content', 'start', 'end'])\n",
    "    # .apply(lambda x: x/100)\n",
    "    .apply(max, axis=1)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54d26eec-1618-4e36-bd76-eaa7ef19f75b",
   "metadata": {},
   "source": [
    "a\n",
    "\n",
    " ### Transformar los datos / Groupby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6e0f713-c7d0-408b-a18e-2d04731b15a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_586709/3511887845.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  disprot.groupby('acc').apply(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "acc\n",
       "B8ZXI1        8.43\n",
       "O08583       10.98\n",
       "O08785        6.67\n",
       "O08807       11.31\n",
       "O08908        6.93\n",
       "             ...  \n",
       "Q9R224      100.00\n",
       "Q9WTL8       21.52\n",
       "Q9WUP7-2      3.05\n",
       "Q9Z0P7       16.74\n",
       "Q9Z1X4       10.69\n",
       "Length: 198, dtype: float64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Agrupar los datos por una columna y aplicar una función a cada grupo\n",
    "# disprot.groupby('acc').size()\n",
    "\n",
    "# disprot.groupby('acc').agg(\n",
    "#   {'disorder_content': ['mean', 'std', 'sum', 'count']}\n",
    "# )\n",
    "\n",
    "# disprot.groupby('acc').apply(\n",
    "#   lambda x: max(x.disorder_content)\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce20dad-a4a1-41b0-ac27-ee6f8d3b0f76",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Concat y Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2c43ba81-0cb7-4321-a28f-2d960d1692ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_586709/4041508714.py:14: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda x: x.iloc[0, :])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>disorder_content</th>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B8ZXI1</td>\n",
       "      <td>8.43</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O08583</td>\n",
       "      <td>10.98</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O08785</td>\n",
       "      <td>6.67</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O08807</td>\n",
       "      <td>11.31</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O08908</td>\n",
       "      <td>6.93</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>O35484</td>\n",
       "      <td>6.70</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      acc  disorder_content  start\n",
       "0  B8ZXI1              8.43    292\n",
       "1  O08583             10.98     77\n",
       "2  O08785              6.67     89\n",
       "3  O08807             11.31    256\n",
       "4  O08908              6.93    423\n",
       "5  O35484              6.70    436"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Concatenar dos dataframes\n",
    "\n",
    "# df1 = disprot.iloc[0:5, :]\n",
    "# df2 = disprot.iloc[20:25, :]\n",
    "\n",
    "# concatenated = pd.concat([df1, df2])\n",
    "# concatenated\n",
    "\n",
    "# # Merge dos dataframes\n",
    "\n",
    "disprot_b = (\n",
    "    disprot\n",
    "        .groupby('acc')\n",
    "        .apply(lambda x: x.iloc[0, :])\n",
    "        .reset_index(drop=True)\n",
    ")\n",
    "disprot_b\n",
    "\n",
    "df1 = disprot_b.loc[0:5, ['acc', 'disorder_content']]\n",
    "df2 = disprot_b.loc[0:5, ['acc', 'start']].sort_values('start')\n",
    "\n",
    "merged = pd.merge(df1, df2, on='acc')\n",
    "merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c9e534c-369c-4ea3-9fac-e7369970ff76",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Pivotear los datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c5b13c77-66a6-4fa5-ad39-9dec7dbe28ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">Value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Property</th>\n",
       "      <th>Dis</th>\n",
       "      <th>Len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ID1</th>\n",
       "      <td>0.6</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID2</th>\n",
       "      <td>0.7</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Value       \n",
       "Property   Dis    Len\n",
       "ID                   \n",
       "ID1        0.6  100.0\n",
       "ID2        0.7  200.0"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "  [\n",
    "    ['ID1', 'Len', 100],\n",
    "    ['ID1', 'Dis', 0.6],\n",
    "    ['ID2', 'Len', 200],\n",
    "    ['ID2', 'Dis', 0.7],\n",
    "  ],\n",
    "  columns=['ID', 'Property', 'Value']\n",
    ")\n",
    "\n",
    "df_pivot = df.pivot(index=[\"ID\"], columns=[\"Property\"], values=[\"Value\"])\n",
    "df_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53979a23-6265-4bdd-8f43-2ee69f028ccd",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Melting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">Value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Property</th>\n",
       "      <th>Dis</th>\n",
       "      <th>Len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ID1</th>\n",
       "      <td>0.6</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID2</th>\n",
       "      <td>0.7</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Value       \n",
       "Property   Dis    Len\n",
       "ID                   \n",
       "ID1        0.6  100.0\n",
       "ID2        0.7  200.0"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "175143e3-347c-48f4-b40b-13fd952068b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>property</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ID1</td>\n",
       "      <td>Dis</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID2</td>\n",
       "      <td>Dis</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID1</td>\n",
       "      <td>Len</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID2</td>\n",
       "      <td>Len</td>\n",
       "      <td>0.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID property  value\n",
       "0  ID1      Dis  100.0\n",
       "1  ID2      Dis  200.0\n",
       "2  ID1      Len    0.6\n",
       "3  ID2      Len    0.7"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "melt_df = (\n",
    "  df_pivot\n",
    "    .reset_index()\n",
    "    .set_axis(\n",
    "      [\"ID\", \"Len\", \"Dis\"],\n",
    "      axis=1\n",
    "    )\n",
    "    .melt(\n",
    "      id_vars=[\"ID\"],\n",
    "      var_name=\"property\",\n",
    "      value_vars=[\"Dis\", \"Len\"]\n",
    "    )\n",
    ")\n",
    "melt_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85019f0b-d013-469b-8571-01449f403eb4",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ## Gráficos con matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8cc14d-5eb8-4b6a-af7c-62652c02887e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a8c07e7-18b2-4985-bdaa-ccad24764c32",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Scatter plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29fd9215-ffae-4784-abf8-47ae70ef737e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "data = pd.read_csv('data/gse260897_counts.csv', sep='\\t')\n",
    "avg_data = (\n",
    "  data\n",
    "    .set_index(\"gene_id\")\n",
    "    .assign(\n",
    "    con = lambda df: df[[\"C1\", \"C2\", \"C3\"]]\n",
    "      .mean(axis=1)\n",
    "      .astype(int),\n",
    "    lip = lambda df: df[[\"L1\", \"L2\", \"L3\"]]\n",
    "      .mean(axis=1)\n",
    "      .astype(int),\n",
    "    )\n",
    "    .filter([\"con\", \"lip\"])\n",
    ")\n",
    "\n",
    "names = pd.read_csv(\"data/gse260897_names.csv\", sep=\"\\t\", header=0)\n",
    "merged_data = pd.merge(\n",
    "  avg_data,\n",
    "  names,\n",
    "  on=\"gene_id\"\n",
    ")\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(10, 5))\n",
    "axes.scatter(\n",
    "  merged_data[\"con\"],\n",
    "  merged_data[\"lip\"],\n",
    "  c=\"red\",\n",
    "  s=10,\n",
    "  alpha=0.5\n",
    ")\n",
    "\n",
    "axes.set_xlabel(\"Mean Raw Counts Control\")\n",
    "axes.set_ylabel(\"Mean Raw Counts Lipid\")\n",
    "axes.set_title(\"Scatter plot of mean raw counts (Control vs Lipid)\")\n",
    "\n",
    "for i, txt in enumerate(merged_data.gene_name):\n",
    "  axes.annotate(\n",
    "    txt,\n",
    "    (merged_data[\"con\"][i], merged_data[\"lip\"][i] + 5000),\n",
    "    ha='center',\n",
    "    va='bottom',\n",
    "    fontsize=12\n",
    "  )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b944d62-7b39-42c1-a0ac-1d7ab5472864",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Line plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d544a4-bed0-4c53-9f8b-90026405f785",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "  ncols=3,\n",
    "  figsize=(15, 5)\n",
    ")\n",
    "for i, (index, row) in enumerate(data.iloc[0:3,:].iterrows()):\n",
    "  axes[i].plot(\n",
    "    [0, 1, 2],\n",
    "    row.loc[[\"C1\", \"C2\", \"C3\"]],\n",
    "    label=\"Control\",\n",
    "    color=\"red\",\n",
    "  )\n",
    "  axes[i].plot(\n",
    "    [0, 1, 2],\n",
    "    row.loc[[\"L1\", \"L2\", \"L3\"]],\n",
    "    label=\"Lipid\",\n",
    "    color=\"blue\",\n",
    "  )\n",
    "  axes[i].set_title(row.gene_id)\n",
    "  axes[i].set_xlabel(\"Sample\")\n",
    "  axes[i].set_ylabel(\"Raw Counts\")\n",
    "  axes[i].legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e17cdc93-583d-4faf-b3a8-155c88281de9",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Bar plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f827a7f9-1580-4452-982b-13ef01b18115",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(15, 5))\n",
    "\n",
    "w = 0.25\n",
    "\n",
    "for i, g in enumerate([\"C1\", \"C2\", \"C3\"]):\n",
    "  axes.bar(\n",
    "    x = [x + i*w for x in range(len(data[g]))],\n",
    "    height=data[g],\n",
    "    label=g,\n",
    "    width=w-0.05,\n",
    "  )\n",
    "axes.legend()\n",
    "axes.set_xticks([x + w for x in range(len(data[g]))])\n",
    "axes.set_xticklabels(data.gene_id, rotation=45)\n",
    "axes.set_xlabel(\"Gene ID\")\n",
    "axes.set_ylabel(\"Raw Counts\")\n",
    "axes.set_title(\"Raw Counts of control samples by gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9913eab-3a44-4c5e-b0ed-ed027c93dba2",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Histograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d99b232-c2e2-4152-a6e4-921ad8254dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "disprot = pd.read_csv('data/DisProt.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "disorders = disprot.groupby('acc').agg(\n",
    "  {'disorder_content': 'max'}\n",
    ")\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(10, 5))\n",
    "\n",
    "axes.hist(\n",
    "  disorders,\n",
    "  bins=20,\n",
    "  color='olive',\n",
    "  edgecolor='black',\n",
    ")\n",
    "axes.set_xlabel(\"Disorder content\")\n",
    "axes.set_ylabel(\"Frequency\")\n",
    "axes.set_title(\"Histogram of disorder content\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ce45362-c67e-431d-93a9-2162a89acea6",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    " ### Box plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6673203a-4bcc-4540-a068-c1af2b5232d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "allp = pd.read_csv(\"data/protein_lengths.csv\", sep=\"\\t\", header=0)\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(10, 5))\n",
    "\n",
    "axes.boxplot(\n",
    "  [\n",
    "    allp.query(\"organism == 'human'\")[\"Length\"],\n",
    "    allp.query(\"organism == 'e_coli'\")[\"Length\"],\n",
    "  ],\n",
    "  showfliers=False\n",
    ")\n",
    "axes.set_xticklabels([\"Human\", \"E. coli\"])\n",
    "axes.set_ylabel(\"Protein length\")\n",
    "axes.set_title(\"Boxplot of protein length by organism\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
