{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "731785c4",
   "metadata": {},
   "source": [
    "# Uso de las bases de datos\n",
    "\n",
    "La interacción con las bases de datos puede hacerse con distintos objetivos.\n",
    "Cada una de estas tareas puede ser realizadas por un bioinformático, pero requiere distintos perfiles de conocimiento:\n",
    "\n",
    "- Realizar un análisis de datos de una vez.\n",
    "- Automatizar un análisis periódico.\n",
    "- Desarrollar una aplicación que utiliza y/o requiere datos a pedido de un usuario.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98ad9d1",
   "metadata": {},
   "source": [
    "## Como obtener datos de las Bases de datos biológicas\n",
    "\n",
    "Las formas más frecuentes de obtener información de las bases de datos son:\n",
    "\n",
    "- Acceso por una página Web:\n",
    "  - Útil para la exploración de pocos datos.\n",
    "  - Interacción fácil con otras bases de datos\n",
    "  - Suele estar bien documentada.\n",
    "  - Puede ser suficiente para obtener datos para hacer un análisis en particular.\n",
    "  - Recuperar grandes cantidades de datos puede ser tedioso y lento.\n",
    "  - No se puede automatizar.\n",
    "\n",
    "- Acceso FTP u otro método de descarga masiva de datos: \n",
    "  - La mayoría de los proveedores permiten la descarga de los datos crudos, en paquetes grandes que contienen la totalidad o grandes porciones de los datos disponibles.\n",
    "  - Suele haber poca documentación.\n",
    "  - Útil para hacer análisis a gran escala, que requiere todos los datos disponibles.\n",
    "  - No permite obtener datos particulares.\n",
    "  - Puede resultar útil si se está desarrollando una aplicación que se actualiza en períodos de tiempo largo.\n",
    "  \n",
    "- Acceso programático:\n",
    "  - Útil para la automatización de tareas y la creación de aplicaciones.\n",
    "  - Poca documentación.\n",
    "  - Restricciones para evitar el abuso.\n",
    "  - Poco eficiente para descarga masiva de datos.\n",
    "  - Requiere más conocimientos informáticos y de programación.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e272496",
   "metadata": {},
   "source": [
    "En esta parte del curso nos vamos en el acceso programático.\n",
    "Para ello, tenemos que entender algunos conceptos básicos de como se transmiten los datos en internet."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11df33ad",
   "metadata": {},
   "source": [
    "## Protocolo HTTP\n",
    "\n",
    "El HTTP (Hypertext Transfer Protocol, Procolo de transferencia de hipertexto) es un protocolo para la comunicación de datos de cualquier medio, de forma distribuida y colaborativa.\n",
    "\n",
    "En este protocolo hay dos actores principales:\n",
    "\n",
    "- El **cliente** (client): Es quien hace la petición de conexión y hace los **pedidos** (request) al servidor.\n",
    "- El **servidor** (server): Es quien recibe los pedidos del cliente y devuelve una **respuesta** (response) a ese pedido.\n",
    "\n",
    "![](https://raw.githubusercontent.com/javieriserte/bases-de-datos-unq-unnoba/main/images/http_scheme.svg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55b505e4",
   "metadata": {},
   "source": [
    "### Principios del protocolo HTTP\n",
    "\n",
    "- Sin conexión permanente:\n",
    "\n",
    "El **cliente** comienza una conexión mediante un **pedido**. El cliente espera al **servidor** que tiene que procesar\n",
    "el **pedido** y luego envía la **respuesta** al **cliente**. Cuando el **cliente** recibe la **respuesta** la conexión termina y tanto el **cliente** y el **servidor** dejan de tener conocimiento de cada uno.\n",
    "\n",
    "- Sin estado:\n",
    "\n",
    "Ni el **cliente** ni el **servidor** pueden mantener información entre distintas conexiones.\n",
    "\n",
    "- No tiene restricciones de los medios que se transfieren.\n",
    "\n",
    "Se puede transferir datos de cualquier tipo, siempre que el **cliente** y el **servidor** sepan como interpretarlos."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd86fc11",
   "metadata": {},
   "source": [
    "## Pedidos y respuestas HTTP\n",
    "\n",
    "Los **pedidos** y las **respuestas** HTTP tienen una estructura fija y contienen datos definidos.\n",
    "\n",
    "### Pedidos HTTP:\n",
    "\n",
    "Los pedidos contienen dos partes.\n",
    "\n",
    "- La primera, llamada **encabezado** (**header**), contiene datos que definen el tipo de pedido y el destinatario.\n",
    "Los elementos más importantes del encabezado son:\n",
    "  - Método: Define el tipo acción que debe realizar el **servidor** al recibir el pedido. Los dos métodos más comunes son:\n",
    "    - GET: Para recuperar información del servidor.\n",
    "    - POST: Para subir datos que requerirá el servidor para procesar el **pedido**.\n",
    "  - Recurso: Ruta del recurso (por ejemplo un archivo al que se dirige el **pedido**).\n",
    "  - Host: Nombre del **servidor**.\n",
    "- La segunda parte son los datos que se envian al servidor. No es obligatoria.\n",
    "\n",
    "<pre>\n",
    "<b style=\"color:red\">POST</b> /<b style=\"color:blue\">Blast.cgi</b> HTTP/2\n",
    "<b style=\"color:green\">Host: blast.ncbi.nlm.nih.gov</b>\n",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0\n",
    "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\n",
    "Accept-Language: es-AR,es;q=0.8,en-US;q=0.5,en;q=0.3\n",
    "Accept-Encoding: gzip, deflate, br\n",
    "Referer: https://blast.ncbi.nlm.nih.gov/Blast.cgi\n",
    "Content-Type: application/x-www-form-urlencoded\n",
    "Content-Length: 1000\n",
    "Origin: https://blast.ncbi.nlm.nih.gov\n",
    "Connection: keep-alive\n",
    "Cookie: ncbi_sid=8D9FBA5\n",
    "Upgrade-Insecure-Requests: 1\n",
    "Sec-GPC: 1\n",
    "\n",
    "<b style=\"color:purple\">ADV_VIEW=on&ALIGNMENTS=100&ALIGNMENT_VIEW=Pairwise</b>\n",
    "</pre>\n",
    "\n",
    "### Respuestas HTTP:\n",
    "\n",
    "Las repuestas tienen un formato similar, pero no van a ser relevantes para nosotros, con exceptión del código de estado.\n",
    "\n",
    "El código de estado (**status code**) es un número que refleja si la consulta fue exitosa o si hubo algún tipo de problema.\n",
    "\n",
    "Algunos de los códigos de estados más comunes son:\n",
    "\n",
    "- 200 OK : La consulta se procesó exitosamente.\n",
    "- 301 Moved Permanently: La url no se encuentra en el lugar indicado, la nueva url se pasa en la respuesta.\n",
    "- 400 Bad Request: El servidor no entiende el pedido.\n",
    "- 404 Not found: La url no se encontró.\n",
    "- 500 Internal Server Error: El servidor encontró una situación que no sabe como manejar."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "557a4391",
   "metadata": {},
   "source": [
    "## REST API\n",
    "\n",
    "Los servidores webs pueden ofrecer sus datos por medio de una interfaz independiente de la representación de una página web. Este tipo de interfaz se conoce como REST-API.\n",
    "\n",
    "- De esta forma es posible interactuar con el servidor de forma más simple.\n",
    "\n",
    "- Esta interfaz está conformada por una colección de **URLs** diferentes.\n",
    "\n",
    "- Cada una de ellas provee una funcionalidad específica.\n",
    "\n",
    "- Por ejemplo, estas son algunas de las **URLs** de NCBI.\n",
    "\n",
    "  - https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\n",
    "  - https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\n",
    "  - https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\n",
    "\n",
    "- Como programadores, podemos interactuar con esta interfaz para llevar las tareas que necesitemos.\n",
    "\n",
    "- La comunicación se realiza por medio de HTTP."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "223c84a4",
   "metadata": {},
   "source": [
    "## Como interactuar con una REST-API de una base de datos.\n",
    "\n",
    "La interacción con una REST-API puede hacerse de muchas formas.\n",
    "\n",
    "Nosotros nos vamos a restringir a dos casos típicos usando el lenguage de programación R."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e22cab1",
   "metadata": {},
   "source": [
    "### Usando la librería httr de R\n",
    "\n",
    "Es un mecanismo básico para hacer peticiones HTTP genéricas.\n",
    "\n",
    "Nosotros tendremos que encargarnos de adaptar los datos a los formatos que sean necesarios y de procesar las respuestas.\n",
    "\n",
    "Suele requerir un manejo más técnico y de más detalles que otras soluciones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "bcb3a96f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Para instalar la libraría\n",
    "# install.packages(\"httr\")\n",
    "\n",
    "library(httr)\n",
    "library(xml2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41c9f9ea",
   "metadata": {},
   "source": [
    "Vamos a comenzar haciendo un pedido con el método GET."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "6179d0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defino la URL\n",
    "url <- \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "e80fcd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# La función GET envía la petición y devuelve la respuesta del servidor\n",
    "response <- GET(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "caf1cc4f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "200"
      ],
      "text/latex": [
       "200"
      ],
      "text/markdown": [
       "200"
      ],
      "text/plain": [
       "[1] 200"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Veamos el status code\n",
    "status_code(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "b0bccf10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n",
      "<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20190110//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190110/einfo.dtd\">\n",
      "<eInfoResult>\n",
      "<DbList>\n",
      "\n",
      "\t<DbName>pubmed</DbName>\n",
      "\t<DbName>protein</DbName>\n",
      "\t<DbName>nuccore</DbName>\n",
      "\t<DbName>ipg</DbName>\n",
      "\t<DbName>nucleotide</DbName>\n",
      "\t<DbName>structure</DbName>\n",
      "\t<DbName>genome</DbName>\n",
      "\t<DbName>annotinfo</DbName>\n",
      "\t<DbName>assembly</DbName>\n",
      "\t<DbName>bioproject</DbName>\n",
      "\t<DbName>biosample</DbName>\n",
      "\t<DbName>blastdbinfo</DbName>\n",
      "\t<DbName>books</DbName>\n",
      "\t<DbName>cdd</DbName>\n",
      "\t<DbName>clinvar</DbName>\n",
      "\t<DbName>gap</DbName>\n",
      "\t<DbName>gapplus</DbName>\n",
      "\t<DbName>grasp</DbName>\n",
      "\t<DbName>dbvar</DbName>\n",
      "\t<DbName>gene</DbName>\n",
      "\t<DbName>gds</DbName>\n",
      "\t<DbName>geoprofiles</DbName>\n",
      "\t<DbName>homologene</DbName>\n",
      "\t<DbName>medgen</DbName>\n",
      "\t<DbName>mesh</DbName>\n",
      "\t<DbName>ncbisearch</DbName>\n",
      "\t<DbName>nlmcatalog</DbName>\n",
      "\t<DbName>omim</DbName>\n",
      "\t<DbName>orgtrack</DbName>\n",
      "\t<DbName>pmc</DbName>\n",
      "\t<DbName>popset</DbName>\n",
      "\t<DbName>proteinclusters</DbName>\n",
      "\t<DbName>pcassay</DbName>\n",
      "\t<DbName>protfam</DbName>\n",
      "\t<DbName>biosystems</DbName>\n",
      "\t<DbName>pccompound</DbName>\n",
      "\t<DbName>pcsubstance</DbName>\n",
      "\t<DbName>seqannot</DbName>\n",
      "\t<DbName>snp</DbName>\n",
      "\t<DbName>sra</DbName>\n",
      "\t<DbName>taxonomy</DbName>\n",
      "\t<DbName>biocollections</DbName>\n",
      "\t<DbName>gtr</DbName>\n",
      "</DbList>\n",
      "\n",
      "</eInfoResult>\n"
     ]
    }
   ],
   "source": [
    "# Veamos el contenido de la respuesta\n",
    "# Primero como texto plano\n",
    "contenido <- content(response, as=\"text\")\n",
    "cat(contenido)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "b8298aba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'list'"
      ],
      "text/latex": [
       "'list'"
      ],
      "text/markdown": [
       "'list'"
      ],
      "text/plain": [
       "[1] \"list\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>$eInfoResult</strong> = <strong>$DbList</strong> = <dl>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'pubmed'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'protein'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'nuccore'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'ipg'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'nucleotide'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'structure'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'genome'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'annotinfo'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'assembly'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'bioproject'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'biosample'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'blastdbinfo'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'books'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'cdd'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'clinvar'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'gap'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'gapplus'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'grasp'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'dbvar'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'gene'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'gds'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'geoprofiles'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'homologene'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'medgen'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'mesh'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'ncbisearch'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'nlmcatalog'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'omim'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'orgtrack'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'pmc'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'popset'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'proteinclusters'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'pcassay'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'protfam'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'biosystems'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'pccompound'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'pcsubstance'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'seqannot'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'snp'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'sra'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'taxonomy'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'biocollections'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$DbName</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'gtr'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\textbf{\\$eInfoResult} = \\textbf{\\$DbList} = \\begin{description}\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'pubmed'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'protein'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'nuccore'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'ipg'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'nucleotide'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'structure'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'genome'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'annotinfo'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'assembly'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'bioproject'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'biosample'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'blastdbinfo'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'books'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'cdd'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'clinvar'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'gap'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'gapplus'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'grasp'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'dbvar'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'gene'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'gds'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'geoprofiles'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'homologene'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'medgen'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'mesh'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'ncbisearch'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'nlmcatalog'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'omim'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'orgtrack'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'pmc'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'popset'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'proteinclusters'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'pcassay'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'protfam'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'biosystems'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'pccompound'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'pcsubstance'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'seqannot'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'snp'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'sra'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'taxonomy'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'biocollections'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$DbName] \\begin{enumerate}\n",
       "\\item 'gtr'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "**$eInfoResult** = **$DbList** = $DbName\n",
       ":   1. 'pubmed'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'protein'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'nuccore'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'ipg'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'nucleotide'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'structure'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'genome'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'annotinfo'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'assembly'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'bioproject'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'biosample'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'blastdbinfo'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'books'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'cdd'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'clinvar'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'gap'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'gapplus'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'grasp'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'dbvar'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'gene'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'gds'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'geoprofiles'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'homologene'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'medgen'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'mesh'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'ncbisearch'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'nlmcatalog'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'omim'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'orgtrack'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'pmc'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'popset'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'proteinclusters'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'pcassay'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'protfam'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'biosystems'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'pccompound'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'pcsubstance'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'seqannot'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'snp'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'sra'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'taxonomy'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'biocollections'\n",
       "\n",
       "\n",
       "\n",
       "$DbName\n",
       ":   1. 'gtr'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$eInfoResult\n",
       "$eInfoResult$DbList\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"pubmed\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"protein\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"nuccore\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"ipg\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"nucleotide\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"structure\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"genome\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"annotinfo\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"assembly\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"bioproject\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"biosample\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"blastdbinfo\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"books\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"cdd\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"clinvar\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"gap\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"gapplus\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"grasp\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"dbvar\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"gene\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"gds\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"geoprofiles\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"homologene\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"medgen\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"mesh\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"ncbisearch\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"nlmcatalog\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"omim\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"orgtrack\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"pmc\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"popset\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"proteinclusters\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"pcassay\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"protfam\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"biosystems\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"pccompound\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"pcsubstance\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"seqannot\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"snp\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"sra\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"taxonomy\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"biocollections\"\n",
       "\n",
       "\n",
       "$eInfoResult$DbList$DbName\n",
       "$eInfoResult$DbList$DbName[[1]]\n",
       "[1] \"gtr\"\n",
       "\n",
       "\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Veamos el contenido de la respuesta\n",
    "# Ahora, transformando el contenido a una lista de R\n",
    "contenido <- as_list(content(response, as=\"parsed\"))\n",
    "class(contenido)\n",
    "contenido"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12dfdf18",
   "metadata": {},
   "source": [
    "Hagamos otra petición, pasando algunos argumentos adicionales."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "e15d835b",
   "metadata": {},
   "outputs": [],
   "source": [
    "url <- \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\"\n",
    "# Según la documentación de EInfo en NCBI, el argumento \"db\" permite obtener detalles\n",
    "# específicos de una base de datos particular.\n",
    "response <- GET(url, query = list(db=\"nuccore\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "3d1c452d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n",
      "<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20190110//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190110/einfo.dtd\">\n",
      "<eInfoResult>\n",
      "\t<DbInfo>\n",
      "\t<DbName>nuccore</DbName>\n",
      "\t<MenuName>Nucleotide</MenuName>\n",
      "\t<Description>Core Nucleotide db</Description>\n",
      "\t<DbBuild>Build210527-2225m.1</DbBuild>\n",
      "\t<Count>467364759</Count>\n",
      "\t<LastUpdate>2021/06/01 22:43</LastUpdate>\n",
      "\t<FieldList>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ALL</Name>\n",
      "\t\t\t<FullName>All Fields</FullName>\n",
      "\t\t\t<Description>All terms from all searchable fields</Description>\n",
      "\t\t\t<TermCount>11591427911</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>UID</Name>\n",
      "\t\t\t<FullName>UID</FullName>\n",
      "\t\t\t<Description>Unique number assigned to each sequence</Description>\n",
      "\t\t\t<TermCount>0</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>Y</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>Y</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>FILT</Name>\n",
      "\t\t\t<FullName>Filter</FullName>\n",
      "\t\t\t<Description>Limits the records</Description>\n",
      "\t\t\t<TermCount>439</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>WORD</Name>\n",
      "\t\t\t<FullName>Text Word</FullName>\n",
      "\t\t\t<Description>Free text associated with record</Description>\n",
      "\t\t\t<TermCount>5493949438</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>TITL</Name>\n",
      "\t\t\t<FullName>Title</FullName>\n",
      "\t\t\t<Description>Words in definition line</Description>\n",
      "\t\t\t<TermCount>356240422</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>KYWD</Name>\n",
      "\t\t\t<FullName>Keyword</FullName>\n",
      "\t\t\t<Description>Nonstandardized terms provided by submitter</Description>\n",
      "\t\t\t<TermCount>20282746</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>AUTH</Name>\n",
      "\t\t\t<FullName>Author</FullName>\n",
      "\t\t\t<Description>Author(s) of publication</Description>\n",
      "\t\t\t<TermCount>3185450</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>JOUR</Name>\n",
      "\t\t\t<FullName>Journal</FullName>\n",
      "\t\t\t<Description>Journal abbreviation of publication</Description>\n",
      "\t\t\t<TermCount>26438</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>VOL</Name>\n",
      "\t\t\t<FullName>Volume</FullName>\n",
      "\t\t\t<Description>Volume number of publication</Description>\n",
      "\t\t\t<TermCount>4074</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ISS</Name>\n",
      "\t\t\t<FullName>Issue</FullName>\n",
      "\t\t\t<Description>Issue number of publication</Description>\n",
      "\t\t\t<TermCount>4577</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PAGE</Name>\n",
      "\t\t\t<FullName>Page Number</FullName>\n",
      "\t\t\t<Description>Page number(s) of publication</Description>\n",
      "\t\t\t<TermCount>184702</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ORGN</Name>\n",
      "\t\t\t<FullName>Organism</FullName>\n",
      "\t\t\t<Description>Scientific and common names of organism, and all higher levels of taxonomy</Description>\n",
      "\t\t\t<TermCount>5521178</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ACCN</Name>\n",
      "\t\t\t<FullName>Accession</FullName>\n",
      "\t\t\t<Description>Accession number of sequence</Description>\n",
      "\t\t\t<TermCount>962879551</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PACC</Name>\n",
      "\t\t\t<FullName>Primary Accession</FullName>\n",
      "\t\t\t<Description>Does not include retired secondary accessions</Description>\n",
      "\t\t\t<TermCount>935456371</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>GENE</Name>\n",
      "\t\t\t<FullName>Gene Name</FullName>\n",
      "\t\t\t<Description>Name of gene associated with sequence</Description>\n",
      "\t\t\t<TermCount>5574349783</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PROT</Name>\n",
      "\t\t\t<FullName>Protein Name</FullName>\n",
      "\t\t\t<Description>Name of protein associated with sequence</Description>\n",
      "\t\t\t<TermCount>7492124</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ECNO</Name>\n",
      "\t\t\t<FullName>EC/RN Number</FullName>\n",
      "\t\t\t<Description>EC number for enzyme or CAS registry number</Description>\n",
      "\t\t\t<TermCount>5957</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PDAT</Name>\n",
      "\t\t\t<FullName>Publication Date</FullName>\n",
      "\t\t\t<Description>Date sequence added to GenBank</Description>\n",
      "\t\t\t<TermCount>11896</TermCount>\n",
      "\t\t\t<IsDate>Y</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>MDAT</Name>\n",
      "\t\t\t<FullName>Modification Date</FullName>\n",
      "\t\t\t<Description>Date of last update</Description>\n",
      "\t\t\t<TermCount>8447</TermCount>\n",
      "\t\t\t<IsDate>Y</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SUBS</Name>\n",
      "\t\t\t<FullName>Substance Name</FullName>\n",
      "\t\t\t<Description>CAS chemical name or MEDLINE Substance Name</Description>\n",
      "\t\t\t<TermCount>178536</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PROP</Name>\n",
      "\t\t\t<FullName>Properties</FullName>\n",
      "\t\t\t<Description>Classification by source qualifiers and molecule type</Description>\n",
      "\t\t\t<TermCount>50852</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SQID</Name>\n",
      "\t\t\t<FullName>SeqID String</FullName>\n",
      "\t\t\t<Description>String identifier for sequence</Description>\n",
      "\t\t\t<TermCount>264366808</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>GPRJ</Name>\n",
      "\t\t\t<FullName>BioProject</FullName>\n",
      "\t\t\t<Description>BioProject</Description>\n",
      "\t\t\t<TermCount>228262</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SLEN</Name>\n",
      "\t\t\t<FullName>Sequence Length</FullName>\n",
      "\t\t\t<Description>Length of sequence</Description>\n",
      "\t\t\t<TermCount>1193293</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>Y</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>FKEY</Name>\n",
      "\t\t\t<FullName>Feature key</FullName>\n",
      "\t\t\t<Description>Feature annotated on sequence</Description>\n",
      "\t\t\t<TermCount>400</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PORG</Name>\n",
      "\t\t\t<FullName>Primary Organism</FullName>\n",
      "\t\t\t<Description>Scientific and common names of primary organism, and all higher levels of taxonomy</Description>\n",
      "\t\t\t<TermCount>5520335</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>COMP</Name>\n",
      "\t\t\t<FullName>Component Accession</FullName>\n",
      "\t\t\t<Description>Component accessions for an assembly</Description>\n",
      "\t\t\t<TermCount>206750026</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ASSM</Name>\n",
      "\t\t\t<FullName>Assembly</FullName>\n",
      "\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t<TermCount>218789</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>DIV</Name>\n",
      "\t\t\t<FullName>Division</FullName>\n",
      "\t\t\t<Description>Division</Description>\n",
      "\t\t\t<TermCount>19</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>STRN</Name>\n",
      "\t\t\t<FullName>Strain</FullName>\n",
      "\t\t\t<Description>Strain</Description>\n",
      "\t\t\t<TermCount>2584762</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ISOL</Name>\n",
      "\t\t\t<FullName>Isolate</FullName>\n",
      "\t\t\t<Description>Isolate</Description>\n",
      "\t\t\t<TermCount>5038345</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>CULT</Name>\n",
      "\t\t\t<FullName>Cultivar</FullName>\n",
      "\t\t\t<Description>Cultivar</Description>\n",
      "\t\t\t<TermCount>38073</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>BRD</Name>\n",
      "\t\t\t<FullName>Breed</FullName>\n",
      "\t\t\t<Description>Breed</Description>\n",
      "\t\t\t<TermCount>5682</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>BIOS</Name>\n",
      "\t\t\t<FullName>BioSample</FullName>\n",
      "\t\t\t<Description>BioSample</Description>\n",
      "\t\t\t<TermCount>1320905</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t</FieldList>\n",
      "\t\t<LinkList>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_assembly</Name>\n",
      "\t\t\t\t<Menu>Assembly</Menu>\n",
      "\t\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t\t<DbTo>assembly</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_assembly_wgscontig</Name>\n",
      "\t\t\t\t<Menu>Assembly</Menu>\n",
      "\t\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t\t<DbTo>assembly</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biocollections</Name>\n",
      "\t\t\t\t<Menu>BioCollections</Menu>\n",
      "\t\t\t\t<Description>BioCollections</Description>\n",
      "\t\t\t\t<DbTo>biocollections</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject</Name>\n",
      "\t\t\t\t<Menu>BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Related BioProject entry</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject_reference</Name>\n",
      "\t\t\t\t<Menu>Reference Genome BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Reference Genome BioProject</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject_repr</Name>\n",
      "\t\t\t\t<Menu>Representative Genome BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Representative Genome BioProject</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biosample</Name>\n",
      "\t\t\t\t<Menu>BioSample links</Menu>\n",
      "\t\t\t\t<Description>BioSample links</Description>\n",
      "\t\t\t\t<DbTo>biosample</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biosystems</Name>\n",
      "\t\t\t\t<Menu>BioSystem Links</Menu>\n",
      "\t\t\t\t<Description>BioSystems</Description>\n",
      "\t\t\t\t<DbTo>biosystems</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_ccds</Name>\n",
      "\t\t\t\t<Menu>ccds</Menu>\n",
      "\t\t\t\t<Description>Link to Consensus CDS</Description>\n",
      "\t\t\t\t<DbTo>ccds</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_clone</Name>\n",
      "\t\t\t\t<Menu>Clone</Menu>\n",
      "\t\t\t\t<Description>Links to Clone</Description>\n",
      "\t\t\t\t<DbTo>clone</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_comp_nuccore</Name>\n",
      "\t\t\t\t<Menu>Component Of</Menu>\n",
      "\t\t\t\t<Description>Link to master record</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_dbvar</Name>\n",
      "\t\t\t\t<Menu>dbVar</Menu>\n",
      "\t\t\t\t<Description>Link from Nucleotide to dbVar</Description>\n",
      "\t\t\t\t<DbTo>dbvar</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_gene</Name>\n",
      "\t\t\t\t<Menu>Gene Links</Menu>\n",
      "\t\t\t\t<Description>Link to related Genes</Description>\n",
      "\t\t\t\t<DbTo>gene</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_genome</Name>\n",
      "\t\t\t\t<Menu>Genome</Menu>\n",
      "\t\t\t\t<Description>Genome record containing nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>genome</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_genome_samespecies</Name>\n",
      "\t\t\t\t<Menu>RefSeq genome for species</Menu>\n",
      "\t\t\t\t<Description>Reference genome or particular genomic segment for the species</Description>\n",
      "\t\t\t\t<DbTo>genome</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_geoprofiles</Name>\n",
      "\t\t\t\t<Menu>GEO Profiles Links</Menu>\n",
      "\t\t\t\t<Description>GEO Profile records associated with Nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>geoprofiles</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_homologene</Name>\n",
      "\t\t\t\t<Menu>HomoloGene Links</Menu>\n",
      "\t\t\t\t<Description>Related HomoloGene</Description>\n",
      "\t\t\t\t<DbTo>homologene</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_mrna_nuccore</Name>\n",
      "\t\t\t\t<Menu>Annotated Genomic</Menu>\n",
      "\t\t\t\t<Description>Genomic sequence records that have the current RNA records as annotated features marking the exons of genes.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_comp</Name>\n",
      "\t\t\t\t<Menu>Component(Core) Links</Menu>\n",
      "\t\t\t\t<Description>Links to Core components</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_gbrs</Name>\n",
      "\t\t\t\t<Menu>Identical RefSeq</Menu>\n",
      "\t\t\t\t<Description>Links GenBank nucleotides to corresponding RefSeq assemblies.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_mgc_refseq</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>NIH cDNA clone</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_mrna</Name>\n",
      "\t\t\t\t<Menu>RNA</Menu>\n",
      "\t\t\t\t<Description>RNA sequence records that are annotated on the current genomic nucleotide sequences. The RNA records represent aligned or predicted transcripts of the current records.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_related</Name>\n",
      "\t\t\t\t<Menu>Genomic Neighbours</Menu>\n",
      "\t\t\t\t<Description>INSDC nucleotides with blast symmetric identity more than 95 percent</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_rsgb</Name>\n",
      "\t\t\t\t<Menu>Identical GenBank</Menu>\n",
      "\t\t\t\t<Description>Links from RefSeq assemblies to source GenBank records.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_samespecies_gbrs</Name>\n",
      "\t\t\t\t<Menu>RefSeq nucleotide for species</Menu>\n",
      "\t\t\t\t<Description>Links from DDBJ/EMBL/GenBank nucleotides or genomic segments to genomic RefSeqs for the same species</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_samespecies_rsgb</Name>\n",
      "\t\t\t\t<Menu>Other nucleotides for species</Menu>\n",
      "\t\t\t\t<Description>Links from genomic RefSeqs to DDBJ/EMBL/GenBank nucleotides or genomic segments for the same species</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_small_genome</Name>\n",
      "\t\t\t\t<Menu>Link to all segments from this genome</Menu>\n",
      "\t\t\t\t<Description>All segments from this genome</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_omim</Name>\n",
      "\t\t\t\t<Menu>OMIM Links</Menu>\n",
      "\t\t\t\t<Description>OMIM records associated with Nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>omim</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pccompound</Name>\n",
      "\t\t\t\t<Menu>PubChem Compound Links</Menu>\n",
      "\t\t\t\t<Description>Related PubChem Compound</Description>\n",
      "\t\t\t\t<DbTo>pccompound</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pcsubstance</Name>\n",
      "\t\t\t\t<Menu>PubChem Substance Links</Menu>\n",
      "\t\t\t\t<Description>Related PubChem Substance</Description>\n",
      "\t\t\t\t<DbTo>pcsubstance</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pmc</Name>\n",
      "\t\t\t\t<Menu>PMC Links</Menu>\n",
      "\t\t\t\t<Description>Free full text articles in PMC</Description>\n",
      "\t\t\t\t<DbTo>pmc</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_popset</Name>\n",
      "\t\t\t\t<Menu>PopSet Links</Menu>\n",
      "\t\t\t\t<Description>Population set containing nucleotide sequences</Description>\n",
      "\t\t\t\t<DbTo>popset</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_probe</Name>\n",
      "\t\t\t\t<Menu>Probe Links</Menu>\n",
      "\t\t\t\t<Description>Related Probe entry</Description>\n",
      "\t\t\t\t<DbTo>probe</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein</Name>\n",
      "\t\t\t\t<Menu>Protein Links</Menu>\n",
      "\t\t\t\t<Description>Protein sequence from Nucleotide coding region</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_mat_peptide</Name>\n",
      "\t\t\t\t<Menu>Mature Peptides</Menu>\n",
      "\t\t\t\t<Description>Link between nucleotide record and mature peptides contained within nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_mgc_url</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Order cDNA clone</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_small_genome</Name>\n",
      "\t\t\t\t<Menu>Link to all proteins from this genome</Menu>\n",
      "\t\t\t\t<Description>All proteins from this genome</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_tsa</Name>\n",
      "\t\t\t\t<Menu>Proteins in this TSA project</Menu>\n",
      "\t\t\t\t<Description>Links to proteins in this TSA project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_tsa_vdb</Name>\n",
      "\t\t\t\t<Menu>Protein from TSA</Menu>\n",
      "\t\t\t\t<Description>Link to proteins in VDB-only TSA project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_wgs</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Links to proteins in this WGS project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_wgs_vdb</Name>\n",
      "\t\t\t\t<Menu>Protein from WGS</Menu>\n",
      "\t\t\t\t<Description>Link to proteins in VDB-only WGS project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_proteinclusters</Name>\n",
      "\t\t\t\t<Menu>Protein Cluster Links</Menu>\n",
      "\t\t\t\t<Description>Related Protein Clusters</Description>\n",
      "\t\t\t\t<DbTo>proteinclusters</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed</Name>\n",
      "\t\t\t\t<Menu>PubMed Links</Menu>\n",
      "\t\t\t\t<Description>PubMed articles cited by Nucleotide sequence record</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed_refseq</Name>\n",
      "\t\t\t\t<Menu>PubMed (RefSeq) Links</Menu>\n",
      "\t\t\t\t<Description>Link to RefSeq PubMed articles</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed_weighted</Name>\n",
      "\t\t\t\t<Menu>PubMed (Weighted) Links</Menu>\n",
      "\t\t\t\t<Description>Links to pubmed</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_snp</Name>\n",
      "\t\t\t\t<Menu>SNP Links</Menu>\n",
      "\t\t\t\t<Description>Related SNPs</Description>\n",
      "\t\t\t\t<DbTo>snp</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sparcle_mrna</Name>\n",
      "\t\t\t\t<Menu>Functional Class</Menu>\n",
      "\t\t\t\t<Description>Functional class of the sequence domain architecture</Description>\n",
      "\t\t\t\t<DbTo>sparcle</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sra</Name>\n",
      "\t\t\t\t<Menu>SRA Links</Menu>\n",
      "\t\t\t\t<Description>SRA Links</Description>\n",
      "\t\t\t\t<DbTo>sra</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sra_wgs</Name>\n",
      "\t\t\t\t<Menu>SRA WGS</Menu>\n",
      "\t\t\t\t<Description>Links to sra</Description>\n",
      "\t\t\t\t<DbTo>sra</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_structure</Name>\n",
      "\t\t\t\t<Menu>Structure Links</Menu>\n",
      "\t\t\t\t<Description>3D structure of Nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>structure</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_taxonomy</Name>\n",
      "\t\t\t\t<Menu>Taxonomy Links</Menu>\n",
      "\t\t\t\t<Description>Taxonomy sequences associated with Nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>taxonomy</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_trace</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Link to related Traces</Description>\n",
      "\t\t\t\t<DbTo>trace</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t</LinkList>\n",
      "\t</DbInfo>\n",
      "\n",
      "</eInfoResult>\n"
     ]
    }
   ],
   "source": [
    "contenido <- content(response, as=\"text\")\n",
    "cat(contenido)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "970ec24c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n",
      "<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20190110//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190110/einfo.dtd\">\n",
      "<eInfoResult>\n",
      "\t<DbInfo>\n",
      "\t<DbName>nuccore</DbName>\n",
      "\t<MenuName>Nucleotide</MenuName>\n",
      "\t<Description>Core Nucleotide db</Description>\n",
      "\t<DbBuild>Build210527-2225m.1</DbBuild>\n",
      "\t<Count>467364759</Count>\n",
      "\t<LastUpdate>2021/06/01 22:43</LastUpdate>\n",
      "\t<FieldList>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ALL</Name>\n",
      "\t\t\t<FullName>All Fields</FullName>\n",
      "\t\t\t<Description>All terms from all searchable fields</Description>\n",
      "\t\t\t<TermCount>11591427911</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>UID</Name>\n",
      "\t\t\t<FullName>UID</FullName>\n",
      "\t\t\t<Description>Unique number assigned to each sequence</Description>\n",
      "\t\t\t<TermCount>0</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>Y</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>Y</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>FILT</Name>\n",
      "\t\t\t<FullName>Filter</FullName>\n",
      "\t\t\t<Description>Limits the records</Description>\n",
      "\t\t\t<TermCount>439</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>WORD</Name>\n",
      "\t\t\t<FullName>Text Word</FullName>\n",
      "\t\t\t<Description>Free text associated with record</Description>\n",
      "\t\t\t<TermCount>5493949438</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>TITL</Name>\n",
      "\t\t\t<FullName>Title</FullName>\n",
      "\t\t\t<Description>Words in definition line</Description>\n",
      "\t\t\t<TermCount>356240422</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>N</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>KYWD</Name>\n",
      "\t\t\t<FullName>Keyword</FullName>\n",
      "\t\t\t<Description>Nonstandardized terms provided by submitter</Description>\n",
      "\t\t\t<TermCount>20282746</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>AUTH</Name>\n",
      "\t\t\t<FullName>Author</FullName>\n",
      "\t\t\t<Description>Author(s) of publication</Description>\n",
      "\t\t\t<TermCount>3185450</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>JOUR</Name>\n",
      "\t\t\t<FullName>Journal</FullName>\n",
      "\t\t\t<Description>Journal abbreviation of publication</Description>\n",
      "\t\t\t<TermCount>26438</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>VOL</Name>\n",
      "\t\t\t<FullName>Volume</FullName>\n",
      "\t\t\t<Description>Volume number of publication</Description>\n",
      "\t\t\t<TermCount>4074</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ISS</Name>\n",
      "\t\t\t<FullName>Issue</FullName>\n",
      "\t\t\t<Description>Issue number of publication</Description>\n",
      "\t\t\t<TermCount>4577</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PAGE</Name>\n",
      "\t\t\t<FullName>Page Number</FullName>\n",
      "\t\t\t<Description>Page number(s) of publication</Description>\n",
      "\t\t\t<TermCount>184702</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ORGN</Name>\n",
      "\t\t\t<FullName>Organism</FullName>\n",
      "\t\t\t<Description>Scientific and common names of organism, and all higher levels of taxonomy</Description>\n",
      "\t\t\t<TermCount>5521178</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ACCN</Name>\n",
      "\t\t\t<FullName>Accession</FullName>\n",
      "\t\t\t<Description>Accession number of sequence</Description>\n",
      "\t\t\t<TermCount>962879551</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PACC</Name>\n",
      "\t\t\t<FullName>Primary Accession</FullName>\n",
      "\t\t\t<Description>Does not include retired secondary accessions</Description>\n",
      "\t\t\t<TermCount>935456371</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>GENE</Name>\n",
      "\t\t\t<FullName>Gene Name</FullName>\n",
      "\t\t\t<Description>Name of gene associated with sequence</Description>\n",
      "\t\t\t<TermCount>5574349783</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PROT</Name>\n",
      "\t\t\t<FullName>Protein Name</FullName>\n",
      "\t\t\t<Description>Name of protein associated with sequence</Description>\n",
      "\t\t\t<TermCount>7492124</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ECNO</Name>\n",
      "\t\t\t<FullName>EC/RN Number</FullName>\n",
      "\t\t\t<Description>EC number for enzyme or CAS registry number</Description>\n",
      "\t\t\t<TermCount>5957</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PDAT</Name>\n",
      "\t\t\t<FullName>Publication Date</FullName>\n",
      "\t\t\t<Description>Date sequence added to GenBank</Description>\n",
      "\t\t\t<TermCount>11896</TermCount>\n",
      "\t\t\t<IsDate>Y</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>MDAT</Name>\n",
      "\t\t\t<FullName>Modification Date</FullName>\n",
      "\t\t\t<Description>Date of last update</Description>\n",
      "\t\t\t<TermCount>8447</TermCount>\n",
      "\t\t\t<IsDate>Y</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SUBS</Name>\n",
      "\t\t\t<FullName>Substance Name</FullName>\n",
      "\t\t\t<Description>CAS chemical name or MEDLINE Substance Name</Description>\n",
      "\t\t\t<TermCount>178536</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PROP</Name>\n",
      "\t\t\t<FullName>Properties</FullName>\n",
      "\t\t\t<Description>Classification by source qualifiers and molecule type</Description>\n",
      "\t\t\t<TermCount>50852</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SQID</Name>\n",
      "\t\t\t<FullName>SeqID String</FullName>\n",
      "\t\t\t<Description>String identifier for sequence</Description>\n",
      "\t\t\t<TermCount>264366808</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>GPRJ</Name>\n",
      "\t\t\t<FullName>BioProject</FullName>\n",
      "\t\t\t<Description>BioProject</Description>\n",
      "\t\t\t<TermCount>228262</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>SLEN</Name>\n",
      "\t\t\t<FullName>Sequence Length</FullName>\n",
      "\t\t\t<Description>Length of sequence</Description>\n",
      "\t\t\t<TermCount>1193293</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>Y</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>FKEY</Name>\n",
      "\t\t\t<FullName>Feature key</FullName>\n",
      "\t\t\t<Description>Feature annotated on sequence</Description>\n",
      "\t\t\t<TermCount>400</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>PORG</Name>\n",
      "\t\t\t<FullName>Primary Organism</FullName>\n",
      "\t\t\t<Description>Scientific and common names of primary organism, and all higher levels of taxonomy</Description>\n",
      "\t\t\t<TermCount>5520335</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>Y</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>COMP</Name>\n",
      "\t\t\t<FullName>Component Accession</FullName>\n",
      "\t\t\t<Description>Component accessions for an assembly</Description>\n",
      "\t\t\t<TermCount>206750026</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ASSM</Name>\n",
      "\t\t\t<FullName>Assembly</FullName>\n",
      "\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t<TermCount>218789</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>DIV</Name>\n",
      "\t\t\t<FullName>Division</FullName>\n",
      "\t\t\t<Description>Division</Description>\n",
      "\t\t\t<TermCount>19</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>STRN</Name>\n",
      "\t\t\t<FullName>Strain</FullName>\n",
      "\t\t\t<Description>Strain</Description>\n",
      "\t\t\t<TermCount>2584762</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>ISOL</Name>\n",
      "\t\t\t<FullName>Isolate</FullName>\n",
      "\t\t\t<Description>Isolate</Description>\n",
      "\t\t\t<TermCount>5038345</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>CULT</Name>\n",
      "\t\t\t<FullName>Cultivar</FullName>\n",
      "\t\t\t<Description>Cultivar</Description>\n",
      "\t\t\t<TermCount>38073</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>BRD</Name>\n",
      "\t\t\t<FullName>Breed</FullName>\n",
      "\t\t\t<Description>Breed</Description>\n",
      "\t\t\t<TermCount>5682</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t\t<Field>\n",
      "\t\t\t<Name>BIOS</Name>\n",
      "\t\t\t<FullName>BioSample</FullName>\n",
      "\t\t\t<Description>BioSample</Description>\n",
      "\t\t\t<TermCount>1320905</TermCount>\n",
      "\t\t\t<IsDate>N</IsDate>\n",
      "\t\t\t<IsNumerical>N</IsNumerical>\n",
      "\t\t\t<SingleToken>Y</SingleToken>\n",
      "\t\t\t<Hierarchy>N</Hierarchy>\n",
      "\t\t\t<IsHidden>N</IsHidden>\n",
      "\t\t</Field>\n",
      "\t</FieldList>\n",
      "\t\t<LinkList>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_assembly</Name>\n",
      "\t\t\t\t<Menu>Assembly</Menu>\n",
      "\t\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t\t<DbTo>assembly</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_assembly_wgscontig</Name>\n",
      "\t\t\t\t<Menu>Assembly</Menu>\n",
      "\t\t\t\t<Description>Assembly</Description>\n",
      "\t\t\t\t<DbTo>assembly</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biocollections</Name>\n",
      "\t\t\t\t<Menu>BioCollections</Menu>\n",
      "\t\t\t\t<Description>BioCollections</Description>\n",
      "\t\t\t\t<DbTo>biocollections</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject</Name>\n",
      "\t\t\t\t<Menu>BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Related BioProject entry</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject_reference</Name>\n",
      "\t\t\t\t<Menu>Reference Genome BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Reference Genome BioProject</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_bioproject_repr</Name>\n",
      "\t\t\t\t<Menu>Representative Genome BioProject Links</Menu>\n",
      "\t\t\t\t<Description>Representative Genome BioProject</Description>\n",
      "\t\t\t\t<DbTo>bioproject</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biosample</Name>\n",
      "\t\t\t\t<Menu>BioSample links</Menu>\n",
      "\t\t\t\t<Description>BioSample links</Description>\n",
      "\t\t\t\t<DbTo>biosample</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_biosystems</Name>\n",
      "\t\t\t\t<Menu>BioSystem Links</Menu>\n",
      "\t\t\t\t<Description>BioSystems</Description>\n",
      "\t\t\t\t<DbTo>biosystems</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_ccds</Name>\n",
      "\t\t\t\t<Menu>ccds</Menu>\n",
      "\t\t\t\t<Description>Link to Consensus CDS</Description>\n",
      "\t\t\t\t<DbTo>ccds</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_clone</Name>\n",
      "\t\t\t\t<Menu>Clone</Menu>\n",
      "\t\t\t\t<Description>Links to Clone</Description>\n",
      "\t\t\t\t<DbTo>clone</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_comp_nuccore</Name>\n",
      "\t\t\t\t<Menu>Component Of</Menu>\n",
      "\t\t\t\t<Description>Link to master record</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_dbvar</Name>\n",
      "\t\t\t\t<Menu>dbVar</Menu>\n",
      "\t\t\t\t<Description>Link from Nucleotide to dbVar</Description>\n",
      "\t\t\t\t<DbTo>dbvar</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_gene</Name>\n",
      "\t\t\t\t<Menu>Gene Links</Menu>\n",
      "\t\t\t\t<Description>Link to related Genes</Description>\n",
      "\t\t\t\t<DbTo>gene</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_genome</Name>\n",
      "\t\t\t\t<Menu>Genome</Menu>\n",
      "\t\t\t\t<Description>Genome record containing nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>genome</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_genome_samespecies</Name>\n",
      "\t\t\t\t<Menu>RefSeq genome for species</Menu>\n",
      "\t\t\t\t<Description>Reference genome or particular genomic segment for the species</Description>\n",
      "\t\t\t\t<DbTo>genome</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_geoprofiles</Name>\n",
      "\t\t\t\t<Menu>GEO Profiles Links</Menu>\n",
      "\t\t\t\t<Description>GEO Profile records associated with Nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>geoprofiles</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_homologene</Name>\n",
      "\t\t\t\t<Menu>HomoloGene Links</Menu>\n",
      "\t\t\t\t<Description>Related HomoloGene</Description>\n",
      "\t\t\t\t<DbTo>homologene</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_mrna_nuccore</Name>\n",
      "\t\t\t\t<Menu>Annotated Genomic</Menu>\n",
      "\t\t\t\t<Description>Genomic sequence records that have the current RNA records as annotated features marking the exons of genes.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_comp</Name>\n",
      "\t\t\t\t<Menu>Component(Core) Links</Menu>\n",
      "\t\t\t\t<Description>Links to Core components</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_gbrs</Name>\n",
      "\t\t\t\t<Menu>Identical RefSeq</Menu>\n",
      "\t\t\t\t<Description>Links GenBank nucleotides to corresponding RefSeq assemblies.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_mgc_refseq</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>NIH cDNA clone</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_mrna</Name>\n",
      "\t\t\t\t<Menu>RNA</Menu>\n",
      "\t\t\t\t<Description>RNA sequence records that are annotated on the current genomic nucleotide sequences. The RNA records represent aligned or predicted transcripts of the current records.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_related</Name>\n",
      "\t\t\t\t<Menu>Genomic Neighbours</Menu>\n",
      "\t\t\t\t<Description>INSDC nucleotides with blast symmetric identity more than 95 percent</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_rsgb</Name>\n",
      "\t\t\t\t<Menu>Identical GenBank</Menu>\n",
      "\t\t\t\t<Description>Links from RefSeq assemblies to source GenBank records.</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_samespecies_gbrs</Name>\n",
      "\t\t\t\t<Menu>RefSeq nucleotide for species</Menu>\n",
      "\t\t\t\t<Description>Links from DDBJ/EMBL/GenBank nucleotides or genomic segments to genomic RefSeqs for the same species</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_samespecies_rsgb</Name>\n",
      "\t\t\t\t<Menu>Other nucleotides for species</Menu>\n",
      "\t\t\t\t<Description>Links from genomic RefSeqs to DDBJ/EMBL/GenBank nucleotides or genomic segments for the same species</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_nuccore_small_genome</Name>\n",
      "\t\t\t\t<Menu>Link to all segments from this genome</Menu>\n",
      "\t\t\t\t<Description>All segments from this genome</Description>\n",
      "\t\t\t\t<DbTo>nuccore</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_omim</Name>\n",
      "\t\t\t\t<Menu>OMIM Links</Menu>\n",
      "\t\t\t\t<Description>OMIM records associated with Nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>omim</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pccompound</Name>\n",
      "\t\t\t\t<Menu>PubChem Compound Links</Menu>\n",
      "\t\t\t\t<Description>Related PubChem Compound</Description>\n",
      "\t\t\t\t<DbTo>pccompound</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pcsubstance</Name>\n",
      "\t\t\t\t<Menu>PubChem Substance Links</Menu>\n",
      "\t\t\t\t<Description>Related PubChem Substance</Description>\n",
      "\t\t\t\t<DbTo>pcsubstance</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pmc</Name>\n",
      "\t\t\t\t<Menu>PMC Links</Menu>\n",
      "\t\t\t\t<Description>Free full text articles in PMC</Description>\n",
      "\t\t\t\t<DbTo>pmc</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_popset</Name>\n",
      "\t\t\t\t<Menu>PopSet Links</Menu>\n",
      "\t\t\t\t<Description>Population set containing nucleotide sequences</Description>\n",
      "\t\t\t\t<DbTo>popset</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_probe</Name>\n",
      "\t\t\t\t<Menu>Probe Links</Menu>\n",
      "\t\t\t\t<Description>Related Probe entry</Description>\n",
      "\t\t\t\t<DbTo>probe</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein</Name>\n",
      "\t\t\t\t<Menu>Protein Links</Menu>\n",
      "\t\t\t\t<Description>Protein sequence from Nucleotide coding region</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_mat_peptide</Name>\n",
      "\t\t\t\t<Menu>Mature Peptides</Menu>\n",
      "\t\t\t\t<Description>Link between nucleotide record and mature peptides contained within nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_mgc_url</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Order cDNA clone</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_small_genome</Name>\n",
      "\t\t\t\t<Menu>Link to all proteins from this genome</Menu>\n",
      "\t\t\t\t<Description>All proteins from this genome</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_tsa</Name>\n",
      "\t\t\t\t<Menu>Proteins in this TSA project</Menu>\n",
      "\t\t\t\t<Description>Links to proteins in this TSA project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_tsa_vdb</Name>\n",
      "\t\t\t\t<Menu>Protein from TSA</Menu>\n",
      "\t\t\t\t<Description>Link to proteins in VDB-only TSA project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_wgs</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Links to proteins in this WGS project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_protein_wgs_vdb</Name>\n",
      "\t\t\t\t<Menu>Protein from WGS</Menu>\n",
      "\t\t\t\t<Description>Link to proteins in VDB-only WGS project</Description>\n",
      "\t\t\t\t<DbTo>protein</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_proteinclusters</Name>\n",
      "\t\t\t\t<Menu>Protein Cluster Links</Menu>\n",
      "\t\t\t\t<Description>Related Protein Clusters</Description>\n",
      "\t\t\t\t<DbTo>proteinclusters</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed</Name>\n",
      "\t\t\t\t<Menu>PubMed Links</Menu>\n",
      "\t\t\t\t<Description>PubMed articles cited by Nucleotide sequence record</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed_refseq</Name>\n",
      "\t\t\t\t<Menu>PubMed (RefSeq) Links</Menu>\n",
      "\t\t\t\t<Description>Link to RefSeq PubMed articles</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_pubmed_weighted</Name>\n",
      "\t\t\t\t<Menu>PubMed (Weighted) Links</Menu>\n",
      "\t\t\t\t<Description>Links to pubmed</Description>\n",
      "\t\t\t\t<DbTo>pubmed</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_snp</Name>\n",
      "\t\t\t\t<Menu>SNP Links</Menu>\n",
      "\t\t\t\t<Description>Related SNPs</Description>\n",
      "\t\t\t\t<DbTo>snp</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sparcle_mrna</Name>\n",
      "\t\t\t\t<Menu>Functional Class</Menu>\n",
      "\t\t\t\t<Description>Functional class of the sequence domain architecture</Description>\n",
      "\t\t\t\t<DbTo>sparcle</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sra</Name>\n",
      "\t\t\t\t<Menu>SRA Links</Menu>\n",
      "\t\t\t\t<Description>SRA Links</Description>\n",
      "\t\t\t\t<DbTo>sra</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_sra_wgs</Name>\n",
      "\t\t\t\t<Menu>SRA WGS</Menu>\n",
      "\t\t\t\t<Description>Links to sra</Description>\n",
      "\t\t\t\t<DbTo>sra</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_structure</Name>\n",
      "\t\t\t\t<Menu>Structure Links</Menu>\n",
      "\t\t\t\t<Description>3D structure of Nucleotide sequence</Description>\n",
      "\t\t\t\t<DbTo>structure</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_taxonomy</Name>\n",
      "\t\t\t\t<Menu>Taxonomy Links</Menu>\n",
      "\t\t\t\t<Description>Taxonomy sequences associated with Nucleotide record</Description>\n",
      "\t\t\t\t<DbTo>taxonomy</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t\t<Link>\n",
      "\t\t\t\t<Name>nuccore_trace</Name>\n",
      "\t\t\t\t<Menu></Menu>\n",
      "\t\t\t\t<Description>Link to related Traces</Description>\n",
      "\t\t\t\t<DbTo>trace</DbTo>\n",
      "\t\t\t</Link>\n",
      "\t\t</LinkList>\n",
      "\t</DbInfo>\n",
      "\n",
      "</eInfoResult>\n"
     ]
    }
   ],
   "source": [
    "# Otra forma de hacerlo es modificar la URL, aunque es más engorroso y es más fácil equivocarnos.\n",
    "\n",
    "url <- \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=nuccore\"\n",
    "response <- GET(url)\n",
    "cat(content(response, as=\"text\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e37cab1",
   "metadata": {},
   "source": [
    "Ahora vamos probar con un POST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "1f6729f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq <- \"MMASTENNEKDNFMRDTASRSKKSRRRSLWIAAGAVPTAIA\\\n",
    "LSLSLASPAAVAQSSFGSSDIIDSGVLDSITRGLTDYLTPR\\\n",
    "DEALPAGEVTYPAIEGLPAGVRVNSAEYVTSHHVVLSIQSA\\\n",
    "AMPERPIKVQLLLPRDWYSSPDRDFPEIWALDGLRAIEKQS\\\n",
    "GWTIETNIEQFFADKNAIVVLPVGGESSFYTDWNEPNNGKN\\\n",
    "YQWETFLTEELAPILDKGFRSNGERAITGISMGGTAAVNIA\\\n",
    "THNPEMFNFVGSFSGYLDTTSNGMPAAIGAALADAGGYNVN\\\n",
    "AMWGPAGSERWLENDPKRNVDQLRGKQVYVSAGSGADDYGQ\\\n",
    "DGSVATGPANAAGVGLELISRMTSQTFVDAANGAGVNVIAN\\\n",
    "FRPSGVHAWPYWQFEMTQAWPYMADSLGMSREDRGADCVAL\\\n",
    "GAIADATADGSLGSCLNNEYLVANGVGRAQDFTNGRAYWSP\\\n",
    "NTGAFGLFGRINARYSELGGPDSWLGFPKTRELSTPDGRGR\\\n",
    "YVHFENGSIYWSAATGPWEIPGDMFTAWGTQGYEAGGLGYP\\\n",
    "VGPAKDFNGGLAQEFQGGYVLRTPQNRAYWVRGAISAKYME\\\n",
    "PGVATTLGFPTGNERLIPGGAFQEFTNGNIYWSASTGAHYI\\\n",
    "LRGGIFDAWGAKGYEQGEYGWPTTDQTSIAAGGETITFQNG\\\n",
    "TIRQVNGRIEESR\"\n",
    "\n",
    "url <- \"https://pfam.xfam.org/search/sequence\"\n",
    "\n",
    "body <- list(\n",
    "    seq = seq,\n",
    "    output=\"xml\"\n",
    ")\n",
    "response <- POST(\n",
    "    url,\n",
    "    body=body\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "e2236c87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
      "<jobs xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n",
      "      xmlns=\"https://pfam.xfam.org/\"\n",
      "      xsi:schemaLocation=\"https://pfam.xfam.org/\n",
      "                          https://pfam.xfam.org/static/documents/schemas/submission.xsd\">\n",
      "  <job job_id=\"9873D490-C4A5-11EB-B6C6-931F0695EB6A\">\n",
      "    <opened>2021-06-03T20:55:07</opened>\n",
      "    <result_url>/search/sequence/resultset/9873D490-C4A5-11EB-B6C6-931F0695EB6A?output=xml</result_url>\n",
      "  </job>\n",
      "</jobs>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cat(content(response, as=\"text\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2154549a",
   "metadata": {},
   "source": [
    "### Usando librerías de terceros específicas para interactuar con alguna base de datos\n",
    "\n",
    "Para muchas bases de datos, existen paquetes que permiten que la interacción son esa base de datos sea más sencilla.\n",
    "\n",
    "Desafortunadamente, no existen para todas las bases de datos y no siempre son estables.\n",
    "En general son desarrollados por una sola persona o equipos pequeños y no siempre están actualizados con\n",
    "respecto a las bases de datos.\n",
    "\n",
    "De todas formas, van a ser nuestra forma preferida de interacción por su sencillez.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "def113b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Para instalar\n",
    "# install.packages(rentrez)\n",
    "\n",
    "library(rentrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6a2e6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "response <- entrez_info()\n",
    "response\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "d598b9dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "response <- entrez_info(db=\"nuccore\")\n",
    "response"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
